docker/Dockerfile

```diff
@@ -5,46 +5,104 @@ ENV LANG en_US.utf8
+RUN apt-get -y install make cmake protobuf-compiler curl bash lld unzip
 
 SHELL ["/bin/bash", "-c"]
 
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --no-modify-path --default-toolchain none -y
+ENV PATH /root/.cargo/bin/:$PATH
+ENV CARGO_INCREMENTAL=0
 
-RUN mkdir -p /risingwave
+COPY rust-toolchain rust-toolchain
 
-WORKDIR /risingwave
+# We need to add the `rustfmt` dependency, otherwise `risingwave_pb` will not compile
+RUN rustup self update \
+  && rustup set profile minimal \
+  && rustup show \
+  && rustup component add rustfmt
+
+RUN cargo install --git https://github.com/xxchan/cargo-chef cargo-chef --locked --rev 11f9fed
 
+FROM rust-base AS rust-planner
+
+RUN mkdir -p /risingwave
+WORKDIR /risingwave
 COPY ./ /risingwave
 
-ENV PATH /root/.cargo/bin/:$PATH
+RUN cargo chef prepare --recipe-path recipe.json
 
-ENV IN_CONTAINER=1
+FROM rust-base AS rust-builder
 
-ARG GIT_SHA
-ENV GIT_SHA=$GIT_SHA
+RUN mkdir -p /risingwave
+WORKDIR /risingwave
 
+COPY --from=rust-planner /risingwave/recipe.json recipe.json
+
+# Build dependencies - this can be cached if the dependencies don't change
+RUN cargo chef cook --release --recipe-path recipe.json
+
+
+# TODO: build dashboard, instead of downloading
 RUN curl -LO https://github.com/risingwavelabs/risingwave/archive/refs/heads/dashboard-artifact.zip
 RUN unzip dashboard-artifact.zip && mv risingwave-dashboard-artifact /risingwave/ui && rm dashboard-artifact.zip
 
-# We need to add the `rustfmt` dependency, otherwise `risingwave_pb` will not compile
-RUN rustup self update \
-  && rustup set profile minimal \
-  && rustup show \
-  && rustup component add rustfmt
 
-RUN cargo fetch && \
-  cargo build -p risingwave_cmd_all --release --features "rw-static-link" && \
-  mkdir -p /risingwave/bin && mv /risingwave/target/release/risingwave /risingwave/bin/ && \
+# Build application
+ARG GIT_SHA
+ENV GIT_SHA=$GIT_SHA
+
+COPY ./ /risingwave
+RUN cargo fetch
+RUN cargo build -p risingwave_cmd_all --release --features "rw-static-link"
+RUN mkdir -p /risingwave/bin && mv /risingwave/target/release/risingwave /risingwave/bin/ && \
   cp ./target/release/build/tikv-jemalloc-sys-*/out/build/bin/jeprof /risingwave/bin/ && \
   chmod +x /risingwave/bin/jeprof && \
   mkdir -p /risingwave/lib && cargo clean
 
-RUN cd /risingwave/java && mvn -B package -Dmaven.test.skip=true -Djava.binding.release=true && \
-    mkdir -p /risingwave/bin/connector-node && \
-    tar -zxvf /risingwave/java/connector-node/assembly/target/risingwave-connector-1.0.0.tar.gz -C /risingwave/bin/connector-node
+# We use rust-maven-plugin to build java-binding. So it's FROM rust-base
+FROM rust-base AS java-builder
+
+RUN apt-get -y install maven
+
+RUN mkdir -p /risingwave
+WORKDIR /risingwave/java
+
+# 1. add pom.xml only here
+# generated via: fd pom.xml | xargs -I {} echo 'COPY {} /risingwave/{}'
+COPY java/common-utils/pom.xml /risingwave/java/common-utils/pom.xml
```

Comment:
Can we move all codes to a temp dir and move all pom file to `/risingwave/java` instead of manually listing all the pom files?